### PR TITLE
#7881: Integrate custom kernels into Mamba SSM block

### PR DESCRIPTION
**Pull request by esmalTT**
This change integrates the `ssm_eltwise_mul` and `ssm_1d_sum_reduce` custom ops into the Mamba model. 

This includes switching out our matmul+mask workaround for `repeat` and `repeat_interleave` to use these new custom ops with masks. Future work will generalize these custom ops so that the `repeat` and `repeat_interleave` ops are no longer required.

### CODE REVIEW

File: models/demos/mamba/tt/mamba_one_step_ssm.py [hunk 1]
```diff
@@ -74,17 +74,14 @@ def preprocess_C(x):
...
         A_weight_name = "mixer.A_log"
 
         def preprocess_A(x):
             x = -torch.exp(x.float())
             # padding with inf
-            # x = F.pad(x, (0, 16), "constant", float("-inf"))
-            x = x.reshape(1, self.hidden_size * self.n)  # (1, 2en)
-            return x.repeat(self.batch_size, 1)  # b, 2en
+            x = torch.nn.functional.pad(x, (0, 16), "constant", float("-inf"))
+            # x = x.reshape(1, self.hidden_size * self.n)  # (1, 2en)
```
Comment (eyonland) [inline models/demos/mamba/tt/mamba_one_step_ssm.py:83]: Do we need to depend on torch within the tt version of Mamba?

File: models/demos/mamba/tt/mamba_one_step_ssm.py [hunk 2]
```diff
@@ -244,45 +209,31 @@ def forward(self, x):
...
-
-        # shard x, C
-        x = ttnn.to_memory_config(x, memory_config=self.configs["sharded_d"])
-        C5 = ttnn.to_memory_config(C4, memory_config=self.configs["sharded_d"])
-        ttnn.deallocate(C4)
-
-        # shard D
-        D = ttnn.to_memory_config(self.D, memory_config=self.configs["sharded_d"])
+        C3 = ttl.operations.primary.transformers.ssm_1d_sum_reduce(
+            C2,
```
Comment (eyonland) [inline models/demos/mamba/tt/mamba_one_step_ssm.py:222]: We need to add ssm_1d_sum_reduce to ttnn.   Ideally we do not use any ttl.operations.

**Review by eyonland — CHANGES_REQUESTED** [review]
We are trying to identify all operations used by Mamba and port them to C++.  Either the operation needs to be ttnn.<operation> or it needs to be ttnn.experimental.

File: models/demos/mamba/tt/mamba_one_step_ssm.py (hunk 1 above)
Comment (esmalTT) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:83]: Unfortunately yes, the slicing and padding we require to preprocess our weights is not provided by `ttnn`

File: models/demos/mamba/tt/mamba_one_step_ssm.py (hunk 2 above)
Comment (esmalTT) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:222]: Should that be done here or in a follow-on PR?

Comment (eyonland) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:222]: Instead of ttl.operations.primary.transformers.ssm_1d_sum_reduce
we need to use 
ttnn.experimental.operations.primary.transformers.ssm_1d_sum_reduce

Does this not work?

File: models/demos/mamba/tt/mamba_one_step_ssm.py (hunk 1 above)
Comment (eyonland) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:83]: Does ttnn.pad not work for you?

Comment (eyonland) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:83]: Can you give me a pytest for this showing it fails?  If so, I'd like to get this fixed asap.

File: models/demos/mamba/tt/mamba_one_step_ssm.py (hunk 2 above)
Comment (esmalTT) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:222]: That works, I'll update the calls to use the ttnn alias

File: models/demos/mamba/tt/mamba_one_step_ssm.py (hunk 1 above)
Comment (esmalTT) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:83]: This code implements the padding part of the preprocessing:
```python
import torch
import ttnn


def test_preprocess_weight(device: ttnn.Device):
    batch = 32
    latent = 16
    hidden_size = 5120
    x = torch.rand((1, 1, hidden_size, latent)) # from state_dict
    x = ttnn.from_torch(
        x, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    y = ttnn.pad(x, ((0, 0), (0, 0), (0, 0), (0, 16)), float("-inf"))
    # y = y[:, :, 0, :].repeat(1, 1, batch, 1)
```

I don't know if the slice (`y = y[:, :, 0, :]`) is supported by ttnn. 

This preprocessing is done at initialization time on the host when we pull the weights from the HF state dictionary.

**Comment by esmalTT** [comment]
@eyonland Do you have any additional comments for this PR?

**Review by arakhmati — APPROVED** [review]

File: models/demos/mamba/tt/mamba_one_step_ssm.py (hunk 1 above)
Comment (eyonland) [reply models/demos/mamba/tt/mamba_one_step_ssm.py:83]: We fallback to torch at the moment to do that sort of slicing.   The intent was to have everying use this slicing and then we will circle back and fix it at some point.

**Review by eyonland — APPROVED** [review]
